File: solutions/python/dnd-character/dnd_character.py
import random

# Constants clearly define the rules of the game
ABILITY_SCORE_SUBTRACT = 10
ABILITY_MODIFIER_DIVISOR = 2
DICE_SIDES = 6
DICE_COUNT_ROLLED = 4


class Character:
    """
    Represents a D&D character with randomly generated ability scores.
    """

    def __init__(self):
        """Initializes a character with random stats."""
        self.strength = ability()
        self.dexterity = ability()
        self.constitution = ability()
        self.intelligence = ability()
        self.wisdom = ability()
        self.charisma = ability()
        self.hitpoints = 10 + modifier(self.constitution)


def ability() -> int:
    """Roll 4d6 and return the sum of the highest 3 dice."""
    rolls = sorted(random.randint(1, DICE_SIDES) for _ in range(DICE_COUNT_ROLLED))
    return sum(rolls[1:])


def modifier(value: int) -> int:
    """Calculate the ability modifier for a given score."""
    return (value - ABILITY_SCORE_SUBTRACT) // ABILITY_MODIFIER_DIVISOR

File: solutions/java/dnd-character/src/main/java/DnDCharacter.java
import java.util.Collection;
import java.util.List;
import java.util.Random;

class DnDCharacter {
    private static final Random random = new Random();

    private final int strength = randomAbility();
    private final int dexterity = randomAbility();
    private final int constitution = randomAbility();
    private final int intelligence = randomAbility();
    private final int wisdom = randomAbility();
    private final int charisma = randomAbility();
    private final int hitpoints = 10 + this.modifier(constitution);

    public static int ability(Collection<Integer> scores) {
        return scores.stream().sorted().skip(1).mapToInt(Integer::intValue).sum();
    }

    private static int randomAbility() {
        return ability(rollDice());
    }

    int modifier(int input) {
        return (int) Math.floor((input - 10) / 2.);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
        return charisma;
    }

    int getHitpoints() {
        return hitpoints;
    }

    public static List<Integer> rollDice() {
        return random.ints(4, 1, 7).boxed().toList();
    }
}
